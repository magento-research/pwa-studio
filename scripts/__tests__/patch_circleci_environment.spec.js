// This tests the script as a JS module (to achieve coverage analysis),
// and also as a separately running process (to mimic CI).
// These scripts aren't run through Babel, so neither are their tests;
// use require() here.
const { exec } = require('child_process');
const { promisify } = require('util');
const { resolve } = require('path');
const mockServer = require('./__helpers__/mock-http-server');

const patchCircleCIEnvironment = require('../patch_circleci_environment');

const execp = promisify(exec);

const patchScript = resolve(__dirname, '../patch_circleci_environment.js');
const nodeBinPath = process.argv[0];

// We must set and mutate these values before mock calls, since
// `mock-http-server` does not support removing handlers that have been added
// with `on`, and we need to test multiple outcomes from the same request.
const mockGithubResponse = jest.fn(() => ({
    status: 400,
    body: 'No mock response was set'
}));
const mockGithubServer = mockServer(mockGithubResponse);
let mockGithubEndpoint;
beforeAll(async () => {
    mockGithubResponse.mockRestore();
    mockGithubEndpoint = await mockGithubServer.mount();
});
afterAll(mockGithubServer.unmount);

beforeEach(() => mockGithubResponse.mockClear());

async function testScript({ env, githubResponse }) {
    if (githubResponse) {
        mockGithubResponse
            .mockReturnValueOnce(githubResponse)
            .mockReturnValueOnce(githubResponse);
    }
    const mockedEnv = Object.assign(
        {
            MOCK_GITHUB_GRAPHQL_ENDPOINT: mockGithubEndpoint
        },
        env
    );

    const mockConfig = {
        env: mockedEnv,
        setEnv: jest.fn(),
        log: jest.fn()
    };

    // test both in-process functionality and child process functionality
    // (for code coverage and CI)

    await patchCircleCIEnvironment(mockConfig);
    const { stdout, stderr, status } = await execp(
        `${nodeBinPath} ${patchScript}`,
        {
            encoding: 'utf8',
            env: mockedEnv
        }
    ).catch(e => e);

    // return a big object for comparison and snapshotting
    return {
        inProcess: {
            stdout: mockConfig.setEnv.mock.calls,
            stderr: mockConfig.log.mock.calls
        },
        childProcess: {
            stderr,
            stdout,
            status
        }
    };
}

test('logs to stderr that PR env vars are already set and exits', async () => {
    const results = await testScript({
        env: {
            CIRCLE_PULL_REQUEST:
                'https://github.com/someOwner/someRepo/pulls/4',
            CIRCLE_PR_USERNAME: 'darkseid'
        }
    });
    expect(results).toMatchSnapshot();
});

test('logs to stderr some PR env vars exist but not all of them', async () => {
    const results = await testScript({
        env: {
            CIRCLE_PROJECT_USERNAME: 'apokolips',
            CIRCLE_PROJECT_REPONAME: 'darkseid',
            CIRCLE_PR_USERNAME: 'new-genesis',
            CIRCLE_PR_REPONAME: 'orion'
        },
        githubResponse: {
            status: 200,
            body: {
                data: {
                    repository: {
                        ref: {
                            associatedPullRequests: {
                                nodes: []
                            }
                        }
                    }
                }
            }
        }
    });
    expect(results).toMatchSnapshot();
    expect(mockGithubResponse.mock.calls[0]).toMatchSnapshot();
});

test('logs to stderr if fetch network-errors', async () => {
    const results = await testScript({
        env: {
            CIRCLE_PROJECT_USERNAME: 'apokolips',
            CIRCLE_PROJECT_REPONAME: 'darkseid',
            MOCK_GITHUB_GRAPHQL_ENDPOINT: 'htllnj8 ynot a real url'
        }
    });
    expect(results).toMatchSnapshot();
});

test('logs to stderr if response comes back non-200', async () => {
    const results = await testScript({
        env: {
            CIRCLE_PROJECT_USERNAME: 'apokolips',
            CIRCLE_PROJECT_REPONAME: 'darkseid',
            CIRCLE_BRANCH: 'missing/branch'
        },
        githubResponse: {
            status: 404,
            body: {
                data: {
                    repository: {
                        ref: {
                            associatedPullRequests: {}
                        }
                    }
                }
            }
        }
    });
    expect(results).toMatchSnapshot();
    expect(mockGithubResponse.mock.calls[0]).toMatchSnapshot();
});

test('writes script that sets env vars to stdout if pr exists', async () => {
    const results = await testScript({
        env: {
            CIRCLE_PROJECT_USERNAME: 'apokolips',
            CIRCLE_PROJECT_REPONAME: 'darkseid',
            CIRCLE_BRANCH: 'missing/branch'
        },
        githubResponse: {
            status: 200,
            body: {
                data: {
                    repository: {
                        ref: {
                            associatedPullRequests: {
                                nodes: [
                                    {
                                        url: 'https://github.com/found/me',
                                        number: 420
                                    }
                                ]
                            }
                        }
                    }
                }
            }
        }
    });
    expect(results).toMatchSnapshot();
    expect(mockGithubResponse.mock.calls[0]).toMatchSnapshot();
});
