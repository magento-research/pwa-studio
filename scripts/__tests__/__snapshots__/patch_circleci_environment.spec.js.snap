// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs to stderr if fetch network-errors 1`] = `
Object {
  "childProcess": Object {
    "status": undefined,
    "stderr": "Failed to contact GitHub API: TypeError: Only absolute URLs are supported
",
    "stdout": "",
  },
  "inProcess": Object {
    "stderr": Array [
      Array [
        "Failed to contact GitHub API: TypeError: Only absolute URLs are supported",
      ],
    ],
    "stdout": Array [],
  },
}
`;

exports[`logs to stderr if response comes back non-200 1`] = `
Object {
  "childProcess": Object {
    "status": undefined,
    "stderr": "Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}

Response: {\\"data\\":{\\"repository\\":{\\"ref\\":{\\"associatedPullRequests\\":{}}}}}

Error: HTTP 404

",
    "stdout": "",
  },
  "inProcess": Object {
    "stderr": Array [
      Array [
        "Failed to find and export pull request variables using the GitHub
API. Check the environment setup step to see what env vars were provided.

Request: {
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}

Response: {\\"data\\":{\\"repository\\":{\\"ref\\":{\\"associatedPullRequests\\":{}}}}}

Error: HTTP 404
",
      ],
    ],
    "stdout": Array [],
  },
}
`;

exports[`logs to stderr if response comes back non-200 2`] = `
Array [
  Object {
    "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}",
    "method": "POST",
  },
]
`;

exports[`logs to stderr some PR env vars exist but not all of them 1`] = `
Object {
  "childProcess": Object {
    "status": undefined,
    "stderr": "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: new-genesis/orion
        instead of:        apokolips/darkseid


Found no pull requests on GitHub for undefined.
",
    "stdout": "",
  },
  "inProcess": Object {
    "stderr": Array [
      Array [
        "Some pull request environment variables are present, and others
are not. DangerCI may not correctly identify this as a pull request. Check the
CircleCI \\"Spin up environment\\" step to see build environment variables.

        Pull request from: new-genesis/orion
        instead of:        apokolips/darkseid

",
      ],
      Array [
        "Found no pull requests on GitHub for undefined.",
      ],
    ],
    "stdout": Array [],
  },
}
`;

exports[`logs to stderr some PR env vars exist but not all of them 2`] = `
Array [
  Object {
    "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/undefined\\"
  }
}",
    "method": "POST",
  },
]
`;

exports[`logs to stderr that PR env vars are already set and exits 1`] = `
Object {
  "childProcess": Object {
    "status": undefined,
    "stderr": "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid
",
    "stdout": "",
  },
  "inProcess": Object {
    "stderr": Array [
      Array [
        "No change: CircleCI has already populated pull request variables
for this build, so 'danger ci' should run as expected.

Pull Request: https://github.com/someOwner/someRepo/pulls/4 from darkseid",
      ],
    ],
    "stdout": Array [],
  },
}
`;

exports[`writes script that sets env vars to stdout if pr exists 1`] = `
Object {
  "childProcess": Object {
    "status": undefined,
    "stderr": "",
    "stdout": "
CI_PULL_REQUEST=\\"https://github.com/found/me\\"
CI_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUEST=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PR_NUMBER=\\"420\\"

",
  },
  "inProcess": Object {
    "stderr": Array [],
    "stdout": Array [
      Array [
        "
CI_PULL_REQUEST=\\"https://github.com/found/me\\"
CI_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUEST=\\"https://github.com/found/me\\"
CIRCLE_PULL_REQUESTS=\\"https://github.com/found/me\\"
CIRCLE_PR_NUMBER=\\"420\\"
",
      ],
    ],
  },
}
`;

exports[`writes script that sets env vars to stdout if pr exists 2`] = `
Array [
  Object {
    "body": "{
  \\"query\\": \\"query getPR($owner: String!, $name: String!, $ref: String!) {\\\\n        repository(owner: $owner, name: $name) {\\\\n          ref(qualifiedName: $ref) {\\\\n            associatedPullRequests(last: 1) {\\\\n              nodes {\\\\n                url\\\\n                number\\\\n              }\\\\n            }\\\\n          }\\\\n        }\\\\n      }\\\\n      \\",
  \\"variables\\": {
    \\"owner\\": \\"apokolips\\",
    \\"name\\": \\"darkseid\\",
    \\"ref\\": \\"refs/heads/missing/branch\\"
  }
}",
    "method": "POST",
  },
]
`;
